tactic/where.lean

```diff
@@ -0,0 +1,226 @@
+| binder_info.implicit := 1
+| binder_info.strict_implicit := 2
+| binder_info.default := 3
+| binder_info.inst_implicit := 4
+| binder_info.aux_decl := 5
+
+meta def binder_less_important (u v : binder_info) : bool :=
+(binder_priority u) < (binder_priority v)
+
+meta def is_in_namespace_nonsynthetic (ns n : name) : bool :=
+ns.is_prefix_of n ∧ ¬(ns.append `user__).is_prefix_of n
+
+meta def get_all_in_namespace (ns : name) : tactic (list name) :=
+do e ← get_env,
+   return $ e.fold [] $ λ d l,
+     if is_in_namespace_nonsynthetic ns d.to_name then d.to_name :: l else l
+
+meta def erase_duplicates {α : Type} [decidable_eq α] : list α → list α
+| [] := []
+| (x :: t) := (x :: erase_duplicates (t.filter $ λ a, a ≠ x))
+
+meta def fetch_potential_variable_names (ns : name) : tactic (list name) :=
+do l ← get_all_in_namespace ns,
+   l ← l.mmap get_def_variables,
+   return $ erase_duplicates $ l.join.map prod.fst
+
+-- Note: we don't call `list.find` and other sorting functions in order
+-- that this file can be imported in places like `data.nat.basic` without
+-- cyclic import hell.
+meta def find_var (n' : name) : list (name × binder_info × expr) → option (name × binder_info × expr)
+| [] := none
+| ((n, bi, e) :: rest) := if n = n' then some (n, bi, e) else find_var rest
+
+meta def is_variable_name (n : name) : lean.parser (option (name × binder_info × expr)) :=
+(do (f, _) ← mk_flag n,
+    l ← get_def_variables f,
+    return $ find_var n l
+) <|> return none
+
+meta def get_variables_core (ns : name) : lean.parser (list (name × binder_info × expr)) :=
+do l ← fetch_potential_variable_names ns,
+   list.filter_map id <$> l.mmap is_variable_name
+
+meta def select_for_which {α β γ : Type} (p : α → β × γ) [decidable_eq β] (b' : β) : list α → list γ × list α
+| [] := ([], [])
+| (a :: rest) :=
+  let (cs, others) := select_for_which rest, (b, c) := p a in
+  if b = b' then (c :: cs, others) else (cs, a :: others)
+
+private meta def collect_by_aux {α β γ : Type} (p : α → β × γ) [decidable_eq β] : list β → list α → list (β × list γ)
+| [] [] := []
+| [] _ := undefined_core "didn't find every key entry!"
+| (b :: rest) as := let (cs, as) := select_for_which p b as in (b, cs) :: collect_by_aux rest as
+
+meta def collect_by {α β γ : Type} (l : list α) (p : α → β × γ) [decidable_eq β] : list (β × list γ) :=
+collect_by_aux p (erase_duplicates $ l.map $ prod.fst ∘ p) l
+
+meta def inflate {α β γ : Type} : list (α × list (β × γ)) → list (α × β × γ)
+| [] := []
+| ((a, l) :: rest) := (l.map $ λ e, (a, e.1, e.2)) ++ inflate rest
+
+private meta def get_least {α : Type} (p : α → α → bool) : α → list α → list α → α × list α
+| c others [] := (c, others)
+| c others (a :: rest) :=
+  if p c a then get_least c (a :: others) rest
+  else get_least a (c :: others) rest
+
+meta def order_by {α : Type} (p : α → α → bool) : list α → list α
+| [] := []
+| (a :: rest) := let (m, others) := get_least p a [] rest in m :: order_by others
```

Comment:
This can be `list.merge_sort` if it is moved to `list.basic.defs`

Reply:
thanks mario